### PR TITLE
Introducing the Articles::Feeds configurations

### PR DESCRIPTION
**Pull request by jeremyf**
## What type of PR is this? (check all applicable)

- [x] Feature

## Description

**tl;dr** This PR is looking to provide a means to programmatically assemble, from system configurations, what is currently the [Articles::Feeds::WeightedQueryStrategy::SCORING_METHOD_CONFIGURATIONS][1].  Once we merge this PR, instead of having that constant, we'll initialize the Articles::Feeds::WeightedQueryStrategy with the variant configuration that we've assembled.

[Loom Walkthrough](https://www.loom.com/share/857cea3698f44a4f876a01fb4e72552c)

**Introduction**

This pull request has quite a bit going on, but as of now the production code does not use any of it.

*Note: None of this code is bleeding into production code paths.*

Put your Ruby hat on and let's go for a ride.  And apologies for not making this a smaller pull request.  As I built this, I made many small commits, but this became the smallest commit that provided the most context without integrating into production code.

Before we get started, you may want to familiarize yourself with where we're going.  The `./app/models/articles/feeds/README.md` provides further guidance; but fair warning reader, I have not reconciled the README's language with what emerged as I wrote this pull request.

Those of you who are part of Forem Core team, you can read the [Refinements to Our Present Feed Configuration][2] to provide some insight into what's happening.

**Why not reconcile?** *Because I want to have all of the language and thoughts available for you to help consider how best to name and model this.*

### On to the Review

The purpose of this pull request is to provide a mechanism for engineering to provide a series of variant query levers.  And to allow us to easily configure those available query levers into a variant query.

**_Why the mix of Ruby configuration and JSON?_**

The Ruby levers are meant to indicate that this is code we don't want to expose in text based configuration because it might create SQL inject points.  (More on that when we integrate the code of this PR into the production implementation)

The JSON levers are meant to indicate that "anyone" can pick the appropriate/available levers and configure how they are set.  In other words, these values do not create SQL injection issues.

### What I Need from You

~~I am putting this forward as a draft so you can read this code ahead of time.  I'm then going to schedule a synchronous code review where we record our collective walk through of the implementation.~~

I have attached added a [Loom Walkthrough](https://www.loom.com/share/857cea3698f44a4f876a01fb4e72552c) that you can watch as part of your review.

[1]:https://github.com/forem/forem/blob/6818ef3ed01b30e4f7651aa0595c8f89aa4055fe/app/services/articles/feeds/weighted_query_strategy.rb#L84-L231
[2]:https://forem.team/jeremy/refinements-to-our-present-feed-configuration-1p0c

### Diagram

Below is a sequence diagram that highlights roughly where I'm planning to go with this work:

A sequence diagram with the Plantuml Description included below

<details>
<summary>PlantUML syntax</summary>

```plantuml
@startuml
!theme amiga

participant "GET relevancy feed" as GetRequest
participant "AbExperiment.get" as AbExperiment
participant "Articles::Feeds::\nVariantQuery.build_for" as VariantQuery
participant "Articles::Feeds::\nVariantAssembler.call" as Assembler
participant "/config/feed-variants/*.json" as VariantConfig
participant "Articles::Feeds\n.lever_catalog" as LeverCatalog

GetRequest --> AbExperiment : with :user
GetRequest <-- AbExperiment : :variant
GetRequest --> VariantQuery : with :user, :variant
VariantQuery --> Assembler : with :variant
Assembler --> VariantConfig : with :variant
Assembler <-- VariantConfig : :variant_config
Assembler --> LeverCatalog : with :variant_config
Assembler <-- LeverCatalog : :query_config
VariantQuery <-- Assembler : :query_config
VariantQuery --> Article : with :user, :query_config
GetRequest <-- Article : Article::ActiveRecord::Relation
@enduml
```
</details>

## Related Tickets & Documents

Closes:

- forem/forem#17268
- forem/forem#17269
- forem/forem#17270
- forem/forem#17271
- forem/forem#17273

Relates to:

- forem/forem#17308
- forem/forem#17245

## QA Instructions, Screenshots, Recordings

Read the code for clarity and questions, we'll pair up to walk through this.

### UI accessibility concerns?

None.

## Added/updated tests?

- [x] Yes

## [Forem core team only] How will this change be communicated?

- [x] I will share this change internally with the appropriate teams

### CODE REVIEW

**Comment by joshpuetz** [comment]
Question not related to the quality of the code: what did you use to create the ERD and sequence diagrams?! They are very helpful!

**Comment by jeremyf** [comment]
I used [PlantUML](http://plantuml.com) which is a wrapper around GraphViz.  Below is the textual representation of that ERD.

```plantuml
@startuml
!theme amiga

entity experiment {
	* id
	--
	* label
	* start date
	* end date
}

entity variant_experiment {
	* experiment_id
	* variant_id
	--
	* probability of using this variant

}

entity variant {
	* id
	--
	* label
	* order_by_lever_id
}

entity variant_relevancy_lever {
	* variant_id
	* relevancy_lever_id
	--
	* config (e.g. relevancy score range)
}

entity order_by_lever {
	* id
	--
	* order_by_fragment
}

entity relevancy_lever {
	* id
	--
	* label
	* user_required
	* select_fragment
	* joins_fragments
	* group_by_fragment
}

experiment ||--|{ variant_experiment
variant ||--o{ variant_experiment
relevancy_lever ||--o{ variant_relevancy_lever
variant ||--o{ variant_relevancy_lever
order_by_lever ||--o{ variant
@enduml
```

File: app/models/articles/feeds/variant_assembler.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      #
+      # @param variant [#to_sym,String,Symbol] the name of the variant we're assembling
+      # @param levers [Articles::Feeds::LeverCatalogBuilder] the available levers for assembly
+      # @param variants [Hash] the cache of previously assembled variants
+      #
+      # @return [Articles::Feeds::VariantQueryConfig]
+      def self.call(variant:, levers: Articles::Feeds.lever_catalog, variants: variants_cache)
+        variant = variant.to_sym
+        variants[variant] ||= begin
+          content = Rails.root.join("config/feed-variants/#{variant}.json").read
```
Comment (joshpuetz) [inline app/models/articles/feeds/variant_assembler.rb:18]: [Suggestion] VERY minor, but it might be worth extracting this path into a constant for easier reference

File: config/feed-variants/20220415-incumbent.json [hunk 2]
```diff
@@ -0,0 +1,100 @@
+{
```
Comment (joshpuetz) [inline config/feed-variants/20220415-incumbent.json:1]: [Suggestion]: It looks like there's some data being conveyed in this file name. Consider a memo or description comment or field inside this json file!

**Review by joshpuetz — APPROVED** [review]

**Review by joshpuetz — COMMENTED** [review]
A few suggestions inline, otherwise this is a FANTASTIC piece of work! 👏🏻

**Comment by msarit** [comment]
@jeremyf Your Loom walkthrough was EXCELLENT Jeremy! I have one question:

does the `relevancy_lever` act on individual posts, or the group of posts that match a `variant_experiment`?

File: app/models/articles/feeds/variant_assembler.rb (hunk 1 above)
Comment (jeremyf) [reply app/models/articles/feeds/variant_assembler.rb:18]: I've been working on the next pull request after this one, and it does include the directory as a constant.  (Which makes it easier to add a directory with broken configurations so we can test the unhappy paths)

File: config/feed-variants/20220415-incumbent.json (hunk 2 above)
Comment (jeremyf) [reply config/feed-variants/20220415-incumbent.json:1]: My original thoughts were to include the name in the file.  However, I'm going with "convention over configuration".  I think a README.md in the config/feed-variants might suffice.

**Comment by jeremyf** [comment]
@msarit I think this may require a sync conversation to work through.

**Comment by jeremyf** [comment]
@msarit when a variant query uses a relevancy lever, it queries all of the articles and determines the relevancy score for that lever.  For example, if the relevancy lever is the `daily_decay` and the variant `daily_decay` configuration cases are `[[0,1],[1,0.5]]` and the fallback is `0`, then any article that is published today will have a daily decay "factor" of 1.  Any article published yesterday (1 day ago) will have a factor of 0.5, and everything else will have a factor of 0.

**Review by msarit — APPROVED** [review]
